### PR TITLE
programs.autorandr: add support for xrandr transformation

### PR DESCRIPTION
**Pull request by kalbasit**
From `man xrandr` (xrandr >= 1.3):

```
       --transform a,b,c,d,e,f,g,h,i
              Specifies a transformation matrix to apply on the output. Automatically a bilinear filter is selected.  The mathematical form corresponds to:
                     a b c
                     d e f
                     g h i
              The transformation is based on homogeneous coordinates. The matrix multiplied by the coordinate vector of a pixel of  the  output  gives  the  transformed coordinate vector of a pixel in the graphic buffer.  More precisely, the vector (x y) of the output pixel is extended to 3 values (x y w), with 1 as the w coordinate and multiplied against the matrix. The final device coordinates of the pixel are then calculated with the so-called homogenic division  by  the transformed w coordinate.  In other words, the device coordinates (x' y') of the transformed pixel are:
                     x' = (ax + by + c) / w'   and
                     y' = (dx + ey + f) / w'   ,
                     with  w' = (gx + hy + i)  .
              Typically,  a and e corresponds to the scaling on the X and Y axes, c and f corresponds to the translation on those axes, and g, h, and i are respectively 0, 0 and 1. The matrix can also be used to express more complex transformations such as keystone correction, or rotation.  For a rotation of an  angle  T, this formula can be used:
                     cos T  -sin T   0
                     sin T   cos T   0
                      0       0      1
              As a special argument, instead of passing a matrix, one can pass the string none, in which case the default values are used (a unit matrix without filter).
```

### CODE REVIEW

File: modules/programs/autorandr.nix [hunk 1]
```diff
@@ -79,6 +79,13 @@ let
         default = null;
         example = "left";
       };
+
+      transform = mkOption {
+        type = types.str;
```
Comment (uvNikita) [inline modules/programs/autorandr.nix]: Maybe it's better to have it as a list of strings? List of floats would be even better, but I guess we might have some weird problems with binary/decimal representations?

Comment (rycee) [reply modules/programs/autorandr.nix]: I think @uvNikita meant to use a list of strings or floats for prettification. Could go further and define a matrix type. I.e., maybe have

```nix
let

  matrixOf = with types;
    let
      list3 = elemType: addCheck (listOf elemType) (l: length l != 3);
    in
      uniq (list3 (list3 float))
      // { description = "3×3 matrix of floating point numbers"; };
```

at the top of the module and use that type here? The example should then become

```nix
[
  [ 0.6 0.0 0.0 ]
  [ 0.0 0.6 0.0 ]
  [ 0.0 0.0 1.0 ]
]
```

and

```nix
lib.concatMapStringsSep "," toString (lib.flatten config.transform)
# Evaluates to
"0.600000,0.000000,0.000000,0.000000,0.600000,0.000000,0.000000,0.000000,1.000000"
```

Note everything above is dry-coded so there might be mistakes.

Comment (kalbasit) [reply modules/programs/autorandr.nix]: @rycee @uvNikita I have updated the PR, thanks for forcing me to learn types! TIL!

File: modules/programs/autorandr.nix [hunk 2]
```diff
@@ -79,6 +79,53 @@ let
         default = null;
         example = "left";
       };
+
+      transform = with types;
+        let
+          transformType = let
+            list3 = elemType: addCheck (listOf elemType) (l: length l == 3);
+          in uniq (list3 (list3 float));
```
Comment (rycee) [inline modules/programs/autorandr.nix]: Why did you remove the `// { description = "3×3 matrix of floating point numbers"; };` part?

File: modules/programs/autorandr.nix [hunk 3]
```diff
@@ -79,6 +79,53 @@ let
...
+        in
+        mkOption {
+          type = nullOr transformType;
+          description  = ''
```
Comment (rycee) [inline modules/programs/autorandr.nix]: Did you have a look at the option description in the generated manual? I think you might want to… 🙂

Comment (rycee) [reply modules/programs/autorandr.nix]: Might be best to simply refer to the xrandr man page.

File: modules/programs/autorandr.nix [hunk 4]
```diff
@@ -79,6 +79,53 @@ let
...
+            i are respectively 0, 0 and 1. The matrix can also be used to
+            express more complex transformations such as keystone correction,
+            or rotation.  For a rotation of an  angle  T, this formula can be
+            used:
+               cos T  -sin T   0
+               sin T   cos T   0
+                0       0      1
+          '';
+          default = null;
+          example = [
```
Comment (rycee) [inline modules/programs/autorandr.nix]: Should be a literal example.

File: modules/programs/autorandr.nix [hunk 5]
```diff
@@ -150,6 +197,7 @@ let
     ${optionalString (config.mode != "") "mode ${config.mode}"}
     ${optionalString (config.rate != "") "rate ${config.rate}"}
     ${optionalString (config.rotate != null) "rotate ${config.rotate}"}
+    ${optionalString (config.transform != null) (concatMapStringsSep "," toString (flatten config.transform))}
```
Comment (rycee) [inline modules/programs/autorandr.nix]: Can break this into two line so it's not so long:
```
    ${optionalString (config.transform != null) (
      concatMapStringsSep "," toString (flatten config.transform)
    )}
```

File: modules/programs/autorandr.nix (hunk 3 above)
Comment (kalbasit) [reply modules/programs/autorandr.nix]: @rycee PTAL.

```
       programs.autorandr.profiles.<name>.config.<name>.transform
           Refer to man 1 xrandr for the documentation of the transform matrix.

           Type: null or 3×3 matrix of floating point numbers

           Default: null

           Example:

               [
                 [ 0.6 0.0 0.0 ]
                 [ 0.0 0.6 0.0 ]
                 [ 0.0 0.0 1.0 ]
               ];

           Declared by:
               <home-manager/modules/programs/autorandr.nix>
```

Comment (rycee) [reply modules/programs/autorandr.nix]: That's nicer! :+1:

You can still improve it by using the docbook method of referencing man pages:

```xml
Refer to 
<citerefentry>
  <refentrytitle>xrandr</refentrytitle>
  <manvolnum>1</manvolnum>
</citerefentry>
for the documentation of the transform matrix.
```

Comment (rycee) [reply modules/programs/autorandr.nix]: Oh and can remove the final `;` in the example to clean it up a little 🙂

Comment (kalbasit) [reply modules/programs/autorandr.nix]: done.

```
       programs.autorandr.profiles.<name>.config.<name>.transform
           Refer to xrandr(1) for the documentation of the transform matrix.

           Type: null or 3×3 matrix of floating point numbers

           Default: null

           Example:

               [
                 [ 0.6 0.0 0.0 ]
                 [ 0.0 0.6 0.0 ]
                 [ 0.0 0.0 1.0 ]
               ]

           Declared by:
               <home-manager/modules/programs/autorandr.nix>

```

**Comment by kalbasit** [comment]
@rycee not sure if you have missed my comment on the resolved discussion but I have addressed your comments.

**Comment by rycee** [comment]
Thanks! Rebased to master in c48fd9d8422519d6f171e7cfc301743d28d08631.

I had to adjust the type a bit since it didn't quite work out with the checks as-is. Maybe in the future can move the `matrixOf` type builder somewhere else for general consumption.
